keps/0031-namespace-management.md

```diff
@@ -0,0 +1,152 @@
+
+### Proposal 1: No namespace creation
+
+All namespace creation is managed outside KUDO. Users such as D2iQ would need to find or build automation for the creation of their namespaces. This REQUIRES that KUDO honor the deployment of resources to the "default" namespace, or to the namespace defined by `--namespace` during install.  
+
+KUDO should further require that NO namespaces be allow in manifests for resources deployed during "install".  The concept of defining manifest which land in multiple namespaces is defined in a latter proposal and is a separate concern. Without multi-namespace support, all resources are install to one namespace and namespace must be created prior to installation and that namespace must be specified during installation.
+
+### Proposal 2: Namespace Creation
+
+KUDO should support the creation of a namespace if it does not exist.  Further KUDO will add a configuration for namespace such that at time of namespace creation it is created with the metadata provided. To this end a manifest file for namespace is to be supported in the template folder:
+
+```
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    ca.istio.io/override: "true"
+    istio-injection: enabled
+    katib-metricscollector-injection: enabled
+  name: {{ .Namespace }}
+```
+
+The template supports the implicit `.Namespace` parameter (which already exists).  The `operator.yaml` file is to will be extended to support `namespaceManifest`.
+
+```
+# operator.yaml
+name: "kubeflow"
+operatorVersion: "0.2.0"
+kudoVersion: "0.10.1"
+kubernetesVersion: 1.15.0
+appVersion: 1.0.0
+namespaceManifest: templates/namespace.yaml
+```
+
+The follow rules apply:
+
+1. Missing `namespaceManifest` means that there is no metadata (creation still applies if warranted)
+1. Missing `name: {{ .Namespace }}` is a failure (and will be validated by package verifier)
+1. No support for static namespace, it must be templated or the manifest should not exist
+1. If no `namespaceManifest` and no `--namespace` the operator installs to "default"
+1. If `--namespace` provided and namespace doesn't exist, KUDO creates it
+1. If KUDO creates namespace, if there is a `namespaceManifest` it uses it to create namespace otherwise does a simple namespace creation.
+1. If KUDO has `namespaceManifest` the check on the namespace is a deepcopy (mean it will ensure that the labels and annotations exists and match) if not it overwrites
+1. if a namespace is created KUDO waits until the namespace is created to move forward in a plan.
+1. if a namespace is needed to be created, it is detected early in the process prior to the deployment plan or as a first step.
+
+#### Alternative Ideas
+
+1. Do we allow for a default namespace other than "default"
+1. We could have another flag to enforce creation, meaning `--namespace` by itself will use that namespace but does NOT create and in combination with `--force` or `--ns-ensure` will create namespace.
```

Comment:
I would prefer this over automatic creation.

Reply:
Yes, I agree, sounds like a good compromise if we choose to support namespace creation.

Reply:
I agree that namespace creation should be **explicit.** However, this won't be enough: namespaces need metadata and so we'll end up having this **and** a `namespaceManifest`. I'd rather choose `namespaceManifest` as it is more flexible and 100% explicit.

Reply:
Well, only in certain cases a namespace needs metadata, right? For just a simple namespace, there wouldn't need to be any metadata and/or manifest. But if there is a manifest, it needs to be applied - either a new namespace needs to be created, or the metadata has to be applied to an existing namespace, correct?

Reply:
I'd rather have **one explicit way** to do it: `namespaceManifest`. The `--namespace` option is only used to populate `name: {{ .Namespace }}`. Also, **no namespace merging:** if there is an existing namespace, we shouldn't alter it: existing namespace should be treated an immutable as changing them has unforeseeable consequences.

Reply:
completely agree with NO merging.. I prefer we do or don't not:)

however I'm now against the `{{ .Namespace` }} templating.  For multiple reasons.  First there will be attempts to use something other than `.Namespace` in the template... or else why would be template it.  The other is explained above... someone might try to `{{ .Namespace }}-extend` which shoud not be allowed.  It seems safest to ensure that KUDO owns what the namespace name is.